python/tvm/relax/training/optimizer.py

```diff
@@ -0,0 +1,577 @@
+            ):
+                with R.dataflow():
+                    x: R.Tensor((3, 3), dtype="float32") = params[0]
+                    x_grad: R.Tensor((3, 3), dtype="float32") = gradients[0]
+                    num_steps: R.Tensor((), dtype="int64") = optim_states[0]
+                    num_steps_new: R.Tensor((), dtype="int64") = R.add(
+                        num_steps, R.const(1, "int64")
+                    )
+                    x_grad_lr: R.Tensor((3, 3), dtype="float32") = R.multiply(
+                        R.const(0.1, "float32"), x_grad
+                    )
+                    x_new: R.Tensor((3, 3), dtype="float32") = R.subtract(x, x_grad_lr)
+                    params_new: R.Tuple(R.Tensor((3, 3), dtype="float32")) = (x_new,)
+                    optim_states_new: R.Tuple(R.Tensor((), dtype="int64")) = (num_steps_new,)
+                    R.output(params_new, optim_states_new)
+                return (params_new, optim_states_new)
+        """
+        raise NotImplementedError()
+
+
+def _get_shape_as_int_list(var):
+    return [int(val) for val in var.struct_info.shape]
+
+
+# We need to subtract on hyperparameters, but do not want to introduce floating point error.
+# Floating point error would lead to a few problems, such as making assert_structural_equal not
+# passed in unit tests
+def _high_precision_subtract(lhs, rhs):
+    return float(Decimal(str(lhs)) - Decimal(str(rhs)))
+
+
+class SGD(Optimizer):
+    """Implements stochastic gradient descent.
+
+    The returned function is equivalent to the following numpy code:
+
+    .. code-block:: python
+        def SGD(param_tuple, grad_tuple, state_tuple):
+            # lr and weight_decay are float constant hyperparameters
+            num_steps = state_tuple[0]
+            state_tuple[0] = num_steps + 1
+            for i in range(len(param_tuple)):
+                param = param_tuple[i]
+                grad = grad_tuple[i]
+                param_tuple[i] = param - lr * (grad + weight_decay * param)
+            return param_tuple, state_tuple
+
+    Parameters
+    ----------
+    lr : float
+        learning rate
+
+    weight_decay : Optional[float]
+        weight decay (L2 penalty) (default: 0)
+    """
+
+    def __init__(self, param_list, lr, weight_decay=0):
+        super().__init__(param_list)
+        self.lr = float(lr)
+        self.weight_decay = float(weight_decay)
+
+    @property
+    def state(self):
+        """The state of SGD is `(num_steps,)`.
+
+        Returns
+        -------
+        res : ADT
+            The state of SGD.
+        """
+        if self._state is None:
+            self._state = tuple_object(
+                (
+                    # num_steps = 0
+                    tvm.nd.array(np.zeros((), "int64")),
+                )
+            )
+        return self._state
+
+    @state.setter
+    def state(self, value):
+        self._state = value
```

Comment:
Curious to know: given that the base class has implemented the setter, do we still have to implement it in subclass?

Reply:
We have to since we override the `@property` method in subclass.

Reply:
I see. Thanks for the explanation.